plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/util/FeatureDevService.kt

```diff
@@ -0,0 +1,243 @@
+
+                if (e is software.amazon.awssdk.services.codewhispererruntime.model.ServiceQuotaExceededException) {
+                    throw MonthlyConversationLimitError(errMssg, e.cause)
+                }
+            }
+            apiError(errMssg, e.cause)
+        } finally {
+            AmazonqTelemetry.startConversationInvoke(
+                amazonqConversationId = conversationId,
+                result = result,
+                reason = failureReason,
+                duration = (System.currentTimeMillis() - startTime).toDouble(),
+                credentialStartUrl = getStartUrl(project = this.project),
+                requestServiceType = "amazoncodewhispererservice"
+            )
+        }
+    }
+
+    fun createUploadUrl(conversationId: String, contentChecksumSha256: String, contentLength: Long):
+        CreateUploadUrlResponse {
+        try {
+            logger.debug { "Executing createUploadUrl with conversationId $conversationId" }
+            val uploadUrlResponse = proxyClient.createTaskAssistUploadUrl(
+                conversationId,
+                contentChecksumSha256,
+                contentLength
+            )
+            logger.debug {
+                "$FEATURE_NAME: Created upload url: {uploadId: ${uploadUrlResponse.uploadId()}, requestId: ${uploadUrlResponse.responseMetadata().requestId()}}"
+            }
+            return uploadUrlResponse
+        } catch (e: Exception) {
+            logger.warn(e) { "$FEATURE_NAME: Failed to generate presigned url: ${e.message}" }
+
+            var errMssg = e.message
+            if (e is CodeWhispererRuntimeException) {
+                errMssg = e.awsErrorDetails().errorMessage()
+                logger.warn(e) { "Create UploadUrl failed for request: ${e.requestId()}" }
+
+                if (e is ValidationException && e.message?.contains("Invalid contentLength") == true) {
+                    throw ContentLengthError(message("amazonqFeatureDev.content_length.error_text"), e.cause)
+                }
+            }
+            apiError(errMssg, e.cause)
+        }
+    }
+
+    suspend fun generatePlan(
+        conversationId: String,
+        uploadId:
+        String,
+        message: String,
+        currentIteration: Int
+    ): GenerateTaskAssistPlanResult {
+        val startTime = System.currentTimeMillis()
+        var failureReason: String? = null
+        var result: Result = Result.Succeeded
+        try {
+            logger.debug { "Executing generateTaskAssistPlan with conversationId $conversationId" }
+            val generatePlanResult = proxyClient.generateTaskAssistPlan(
+                conversationId,
+                uploadId,
+                message
+            )
+            return generatePlanResult
+        } catch (e: Exception) {
+            logger.warn(e) { "$FEATURE_NAME: Failed to execute planning : ${e.message}" }
+            failureReason = e.javaClass.simpleName
```

Comment:
Should this `failureReason` be updated based on specific exception handling for ServiceQuotaExceededException ?

Reply:
I think for VSCode we explicitly show `PlanIterationLimitError`; but we could keep it as is here given the `ServiceQuotaExceededException` will be specific for different APIs like createConversation, planning and codegen